store/store_test.go

```diff
@@ -37,56 +37,101 @@ func initStore(t *testing.T, dir string) (*store.Store, error) {
+		s, err := initStore(t, tempDir, false)
+		require.NoError(t, err)
+		blks := testutil.GenerateBlocksOfSize(2, 100)
+
+		t.Logf("Putting a new block")
+		err = s.Put(blks[0].Cid().Bytes(), blks[0].RawData())
+		require.NoError(t, err)
+		value, found, err := s.Get(blks[0].Cid().Bytes())
+		require.NoError(t, err)
+		require.True(t, found)
+		require.Equal(t, value, blks[0].RawData())
+
+		t.Logf("Overwrite same key with different value")
+		err = s.Put(blks[0].Cid().Bytes(), blks[1].RawData())
+		require.NoError(t, err)
+		value, found, err = s.Get(blks[0].Cid().Bytes())
+		require.NoError(t, err)
+		require.True(t, found)
+		require.Equal(t, value, blks[1].RawData())
+
+		t.Logf("Overwrite same key with same value")
+		err = s.Put(blks[0].Cid().Bytes(), blks[1].RawData())
+		require.Error(t, err, types.ErrKeyExists.Error())
+		value, found, err = s.Get(blks[0].Cid().Bytes())
+		require.NoError(t, err)
+		require.True(t, found)
+		require.Equal(t, value, blks[1].RawData())
+
+		s.Flush()
+
+		// Start iterator
+		flPath := filepath.Join(tempDir, "storethehash.index.free")
+		file, err := os.Open(flPath)
+		require.NoError(t, err)
+		t.Cleanup(func() { require.NoError(t, file.Close()) })
+
+		iter := freelist.NewFreeListIter(file)
+		// Check freelist for the only update. Should be the first position
+		blk, err := iter.Next()
+		require.Equal(t, blk.Offset, types.Position(0))
+		require.NoError(t, err)
+		// Check that is the last
+		_, err = iter.Next()
+		require.EqualError(t, err, io.EOF.Error())
+	})
+	t.Run("when immutable", func(t *testing.T) {
+		tempDir := t.TempDir()
+		s, err := initStore(t, tempDir, true)
+		require.NoError(t, err)
+		blks := testutil.GenerateBlocksOfSize(2, 100)
+
+		t.Logf("Putting a new block")
+		err = s.Put(blks[0].Cid().Bytes(), blks[0].RawData())
+		require.NoError(t, err)
+		value, found, err := s.Get(blks[0].Cid().Bytes())
+		require.NoError(t, err)
+		require.True(t, found)
+		require.Equal(t, value, blks[0].RawData())
+
+		t.Logf("Overwrite same key with different value")
+		err = s.Put(blks[0].Cid().Bytes(), blks[1].RawData())
+		require.Error(t, err, types.ErrKeyExists.Error())
+		value, found, err = s.Get(blks[0].Cid().Bytes())
+		require.NoError(t, err)
+		require.True(t, found)
+		require.Equal(t, value, blks[0].RawData())
+
+		t.Logf("Overwrite same key with same value")
+		err = s.Put(blks[0].Cid().Bytes(), blks[1].RawData())
+		require.Error(t, err, types.ErrKeyExists.Error())
+		value, found, err = s.Get(blks[0].Cid().Bytes())
+		require.NoError(t, err)
+		require.True(t, found)
+		require.Equal(t, value, blks[0].RawData())
+
+		s.Flush()
+
+		// Start iterator
+		flPath := filepath.Join(tempDir, "storethehash.index.free")
+		file, err := os.Open(flPath)
+		require.NoError(t, err)
+		t.Cleanup(func() { require.NoError(t, file.Close()) })
+
+		iter := freelist.NewFreeListIter(file)
+		// Check freelist -- no updates
+		_, err = iter.Next()
+		require.EqualError(t, err, io.EOF.Error())
+	})
```

Comment:
Include in this test two keys that have the same bit prefix to test my suggested changes.

Reply:
will do to test my hypothesis about getPrimaryKeyData